File: tests/unit/native/test_shell_safety.cpp
// Test for src/runtime/tools/shell_safety.h (plan: commit 0582e09 "Study from
// hermes" -- file/bash/safety.py + file/bash/output_enhance.py).
// This test covers:
// - command_detection_variants: collapse/dedup/deobfuscate
// - detect_hardline_command: all 7 ordered checks with reference goldens
// - check_hardline_blocked: obfuscation defeated via variants
// - foreground_background_guidance: 12 patterns + quoted-span stripping
// - base_command_name: first non-assignment command word
// - annotate_failure: 4000-char sample, module-not-found capture

#include "ut/ut.hpp"
#include "bench_util.h"
#include <runtime/tools/shell_safety.h>

#include <cstdio>
#include <string>

using namespace boost::ut;
using namespace boost::ut::literals;
using namespace kimix::runtime::tools;

namespace {
kimix::string_view sv(const std::string& s) { return kimix::string_view(s); }

kimix::vector<kimix::string> variants_of(const char* cmd) {
    kimix::vector<kimix::string> out;
    command_detection_variants(sv(cmd), out);
    return out;
}

const char* kFgBg = "Long-running process detected. Consider mode='send' "
                    "(background) + TaskOutput to avoid blocking on timeout.";

// 10k diverse commands (safe + blocked + obfuscated mixes) with per-template
// expectations: <count> commands, <variants> total detection variants,
// <detect> blocked by detect_hardline_command, <check> blocked by
// check_hardline_blocked.  The templates mirror the unit-test goldens above,
// so a semantics break shows up as a count mismatch.
struct cmd_fixtures {
    kimix::vector<kimix::string> cmds;
    size_t e_variants = 0;
    size_t e_detect = 0;
    size_t e_check = 0;
    double bytes = 0.0;
};

cmd_fixtures build_cmd_fixtures() {
    cmd_fixtures f;
    f.cmds.reserve(10000);
    char buf[128];
    const size_t kRm = 2000, kSafe = 2000, kObf = 1000, kMkfs = 1000,
                 kDd = 1000, kShut = 1000, kKill = 500, kKillSafe = 500,
                 kFmt = 500, kGit = 500;
    for (size_t i = 0; i < kRm; ++i) {
        std::snprintf(buf, sizeof(buf), "rm -rf / ; echo %zu", i);
        f.cmds.emplace_back(buf);
    }
    for (size_t i = 0; i < kSafe; ++i) {
        std::snprintf(buf, sizeof(buf), "echo hello world %zu", i);
        f.cmds.emplace_back(buf);
    }
    for (size_t i = 0; i < kObf; ++i) {
        std::snprintf(buf, sizeof(buf), "r\\m -rf / ; echo x %zu", i);
        f.cmds.emplace_back(buf);
    }
    for (size_t i = 0; i < kMkfs; ++i) {
        std::snprintf(buf, sizeof(buf), "mkfs.ext4 /dev/sda%zu", i % 4);
        f.cmds.emplace_back(buf);
    }
    for (size_t i = 0; i < kDd; ++i) {
        std::snprintf(buf, sizeof(buf), "dd if=/dev/zero of=/dev/sda%zu",
                      i % 4);
        f.cmds.emplace_back(buf);
    }
    for (size_t i = 0; i < kShut; ++i) {
        std::snprintf(buf, sizeof(buf), "'shutdown' -h now ; echo %zu", i);
        f.cmds.emplace_back(buf);
    }
    for (size_t i = 0; i < kKill; ++i) {
        std::snprintf(buf, sizeof(buf), "kill -9 1 ; echo %zu", i);
        f.cmds.emplace_back(buf);
    }
    for (size_t i = 0; i < kKillSafe; ++i) {
        std::snprintf(buf, sizeof(buf), "kill 123");
        f.cmds.emplace_back(buf);
    }
    for (size_t i = 0; i < kFmt; ++i) {
        std::snprintf(buf, sizeof(buf), "format C: ; echo %zu", i);
        f.cmds.emplace_back(buf);
    }
    for (size_t i = 0; i < kGit; ++i) {
        std::snprintf(buf, sizeof(buf), "git status && grep foo %zu", i);
        f.cmds.emplace_back(buf);
    }
    f.e_variants = kRm * 1 + kSafe * 1 + kObf * 2 + kMkfs * 1 + kDd * 1 +
                   kShut * 2 + kKill * 1 + kKillSafe * 1 + kFmt * 2 + kGit * 1;
    f.e_detect = kRm + kMkfs + kDd + kKill + kFmt;
    f.e_check = kRm + kObf + kMkfs + kDd + kShut + kKill + kFmt;
    for (const auto& c : f.cmds) {
        f.bytes += static_cast<double>(c.size());
    }
    return f;
}
} // namespace

int main(int argc, char* argv[]) {
    boost::ut::detail::cfg::parse_arg_with_fallback(
        argc, const_cast<const char**>(argv));

    "variants_basic"_test = [] {
        auto v = variants_of("rm -rf /");
        expect((v.size() == 1u));
        expect((v[0] == "rm -rf /"));
        v = variants_of("  rm   -rf  /  ");
        expect((v.size() == 1u));
        expect((v[0] == "rm -rf /"));
        v = variants_of("");
        expect(v.empty());
        v = variants_of("   ");
        expect(v.empty());
    };

    "variants_deobfuscate"_test = [] {
        auto v = variants_of("r\\m -rf /");
        expect((v.size() == 2u));
        expect((v[0] == "r\\m -rf /"));
        expect((v[1] == "rm -rf /"));
        v = variants_of("\"rm\" -rf /");
        expect((v.size() == 2u));
        expect((v[0] == "\"rm\" -rf /"));
        expect((v[1] == "rm -rf /"));
        v = variants_of("RM -RF /");
        expect((v.size() == 2u));
        expect((v[0] == "RM -RF /"));
        expect((v[1] == "rm -rf /"));
        // quoted span: original, dequoted-lowered, lowered all distinct
        v = variants_of("echo 'hi'");
        expect((v.size() == 2u));
        expect((v[0] == "echo 'hi'"));
        expect((v[1] == "echo hi"));
    };

    "hardline_recursive_delete"_test = [] {
        hardline_result r = detect_hardline_command(sv("rm -rf /"));
        expect(r.blocked);
        expect((*r.description == "Recursive delete of protected root/home (`/`)"));
        r = detect_hardline_command(sv("rm -rf ~"));
        expect(r.blocked);
        expect((*r.description == "Recursive delete of protected root/home (`~`)"));
        r = detect_hardline_command(sv("rm -rf $HOME"));
        expect(r.blocked);
        expect((*r.description == "Recursive delete of protected root/home (`$home`)"));
        r = detect_hardline_command(sv("rm -rf ${HOME}"));
        expect(r.blocked);
        // the message names the RAW lowercased token (${home}), not $home
        expect((*r.description == "Recursive delete of protected root/home (`${home}`)"));
        r = detect_hardline_command(sv("rm -rf /*"));
        expect(r.blocked);
        r = detect_hardline_command(sv("rm -r /"));
        expect(r.blocked);
        r = detect_hardline_command(sv("rm -f /"));
        expect(r.blocked);
        // no r/f flags -> not armed
        r = detect_hardline_command(sv("rm /"));
        expect(!r.blocked);
        // deeper paths are never protected
        r = detect_hardline_command(sv("rm -rf /tmp/build"));
        expect(!r.blocked);
        // rmdir needs r/s
        r = detect_hardline_command(sv("rmdir /"));
        expect(!r.blocked);
        r = detect_hardline_command(sv("rmdir -r /"));
        expect(r.blocked);
        r = detect_hardline_command(sv("rmdir -s /"));
        expect(r.blocked);
        r = detect_hardline_command(sv("rmdir -p /"));
        expect(!r.blocked);
        // Windows drive roots
        r = detect_hardline_command(sv("rm -rf C:\\"));
        expect(r.blocked);
        expect((*r.description == "Recursive delete of protected root/home (`c:\\`)"));
        r = detect_hardline_command(sv("rm -rf c:/"));
        expect(r.blocked);
        r = detect_hardline_command(sv("del /f /s /q C:\\*"));
        expect(r.blocked);
        expect((*r.description == "Recursive delete of protected root/home (`c:\\*`)"));
        r = detect_hardline_command(sv("del /q C:\\Windows"));
        expect(!r.blocked);
        // rmdir vs rm alternation: "rmdir" matches the longer alternative
        r = detect_hardline_command(sv("rmdir -rf /"));
        expect(r.blocked);
    };

    "hardline_other_rules"_test = [] {
        // mkfs
        hardline_result r = detect_hardline_command(sv("mkfs.ext4 /dev/sda1"));
        expect(r.blocked);
        expect((*r.description == "Disk formatting command (`mkfs`) is blocked"));
        r = detect_hardline_command(sv("mkfs.fat -n USB /dev/sdb1"));
        expect(r.blocked);
        // dd to raw device
        r = detect_hardline_command(sv("dd if=/dev/zero of=/dev/sda"));
        expect(r.blocked);
        expect((*r.description == "`dd` writing to a raw device is blocked"));
        r = detect_hardline_command(sv("dd of=/dev/rdisk1"));
        expect(r.blocked);
        r = detect_hardline_command(sv("dd if=/dev/zero of=/tmp/out"));
        expect(!r.blocked);
        // power commands (first token)
        r = detect_hardline_command(sv("shutdown -h now"));
        expect(r.blocked);
        expect((*r.description == "System `shutdown` command is blocked"));
        r = detect_hardline_command(sv("reboot"));
        expect(r.blocked);
        r = detect_hardline_command(sv("poweroff"));
        expect(r.blocked);
        r = detect_hardline_command(sv("halt"));
        expect(r.blocked);
        // fork bomb
        r = detect_hardline_command(sv(":(){ :|:& };:"));
        expect(r.blocked);
        expect((*r.description == "Fork bomb pattern detected"));
        r = detect_hardline_command(sv(":(){ :& };:"));
        expect(!r.blocked);
        // kill PID 1 / $PPID
        r = detect_hardline_command(sv("kill 1"));
        expect(r.blocked);
        expect((*r.description == "`kill` targeting PID 1 (or `$PPID`) is blocked"));
        r = detect_hardline_command(sv("kill -9 1"));
        expect(r.blocked);
        r = detect_hardline_command(sv("kill $PPID"));
        expect(r.blocked);
        r = detect_hardline_command(sv("kill 123"));
        expect(!r.blocked);
        r = detect_hardline_command(sv("killall 1"));
        expect(!r.blocked);
        // Windows format
        r = detect_hardline_command(sv("format C:"));
        expect(r.blocked);
        expect((*r.description == "Windows `format` on a drive is blocked"));
        r = detect_hardline_command(sv("format D:\\"));
        expect(r.blocked);
        r = detect_hardline_command(sv("format C:\\Windows"));
        expect(!r.blocked);
        // benign
        r = detect_hardline_command(sv("ls -la"));
        expect(!r.blocked);
        r = detect_hardline_command(sv(""));
        expect(!r.blocked);
        r = detect_hardline_command(sv("   "));
        expect(!r.blocked);
    };

    "hardline_detect_does_not_deobfuscate"_test = [] {
        // detect_hardline_command only collapses whitespace; quote/backslash
        // tricks need check_hardline_blocked's variants.
        expect(!detect_hardline_command(sv("r\\m -rf /")).blocked);
        expect(!detect_hardline_command(sv("r'm' -rf /")).blocked);
        expect(!detect_hardline_command(sv("rm '-rf' /")).blocked);
    };

    "check_hardline_blocked"_test = [] {
        hardline_result r = check_hardline_blocked(sv("r\\m -rf /"));
        expect(r.blocked);
        expect((*r.description == "Recursive delete of protected root/home (`/`)"));
        r = check_hardline_blocked(sv("r'm' -rf /"));
        expect(r.blocked);
        r = check_hardline_blocked(sv("rm '-rf' /"));
        expect(r.blocked);
        r = check_hardline_blocked(sv("echo hello"));
        expect(!r.blocked);
        r = check_hardline_blocked(sv(""));
        expect(!r.blocked);
        r = check_hardline_blocked(sv("sudo rm -rf /"));
        expect(r.blocked);
    };

    "guidance_patterns"_test = [] {
        expect((*foreground_background_guidance(sv("npm run dev")) == kFgBg));
        expect((*foreground_background_guidance(sv("npm run start")) == kFgBg));
        expect(!foreground_background_guidance(sv("npm run build")).has_value());
        expect((*foreground_background_guidance(sv("pnpm run watch")) == kFgBg));
        expect((*foreground_background_guidance(sv("yarn run serve")) == kFgBg));
        expect((*foreground_background_guidance(sv("bun run dev")) == kFgBg));
        expect((*foreground_background_guidance(sv("next dev")) == kFgBg));
        expect((*foreground_background_guidance(sv("vite")) == kFgBg));
        expect((*foreground_background_guidance(sv("nodemon app.js")) == kFgBg));
        expect((*foreground_background_guidance(sv("uvicorn app:app")) == kFgBg));
        expect((*foreground_background_guidance(sv("gunicorn app:app")) == kFgBg));
        expect((*foreground_background_guidance(sv("python -m http.server")) == kFgBg));
        expect((*foreground_background_guidance(sv("docker compose up")) == kFgBg));
        expect((*foreground_background_guidance(sv("docker-compose up")) == kFgBg));
        expect((*foreground_background_guidance(sv("echo done &")) == kFgBg));
        expect((*foreground_background_guidance(sv("nohup python app.py")) == kFgBg));
        expect((*foreground_background_guidance(sv("setsid bash")) == kFgBg));
        // keywords inside quoted spans are ignored
        expect(!foreground_background_guidance(sv("echo 'npm run dev'")).has_value());
        expect(!foreground_background_guidance(sv("echo \"docker compose up\"")).has_value());
        // benign / empty
        expect(!foreground_background_guidance(sv("ls -la")).has_value());
        expect(!foreground_background_guidance(sv("npmrun dev")).has_value());
        expect(!foreground_background_guidance(sv("")).has_value());
        expect(!foreground_background_guidance(sv("   ")).has_value());
    };

    "base_command_name"_test = [] {
        expect((base_command_name(sv("/usr/bin/grep -r foo")) == "grep"));
        expect((base_command_name(sv("FOO=1 git diff")) == "git"));
        expect((base_command_name(sv("python -m http.server")) == "python"));
        expect((base_command_name(sv("app.exe --help")) == "app"));
        expect((base_command_name(sv("cmd.exe")) == "cmd"));
        expect((base_command_name(sv("echo hi && grep foo")) == "grep"));
        expect((base_command_name(sv("ls")) == "ls"));
        expect((base_command_name(sv("/x/y/tool.EXE arg")) == "tool"));
        expect((base_command_name(sv("python3.11 -V")) == "python3.11"));
        expect((base_command_name(sv("a=b c=d")) == ""));
        expect((base_command_name(sv("")) == ""));
        expect((base_command_name(sv("   ")) == ""));
    };

    "annotate_failure"_test = [] {
        expect(!annotate_failure(sv(""), sv("x"), 1).has_value());
        auto r = annotate_failure(sv("bash: foo: command not found"), sv("foo"), 127);
        expect(r.has_value());
        expect((*r == "The command was not found. Check it is installed and on "
                      "PATH (use `which <cmd>` / `Get-Command <cmd>`)."));
        r = annotate_failure(sv("'foo' is not recognized as an internal or "
                                "external command"), sv("foo"), 1);
        expect(r.has_value());
        expect((*r == "The command was not found. Check it is installed and on "
                      "PATH (use `which <cmd>` / `Get-Command <cmd>`)."));
        r = annotate_failure(sv("ls: cannot access 'x': No such file or "
                                "directory"), sv("ls"), 2);
        expect((*r == "A file or directory referenced by the command does not "
                      "exist. Verify the path with `Glob`/ReadFile."));
        r = annotate_failure(sv("ModuleNotFoundError: No module named "
                                "'requests'"), sv("python"), 1);
        expect((*r == "Python module requests is missing. Install it (e.g. "
                      "`pip install requests`) or check the environment."));
        r = annotate_failure(sv("MODULENOTFOUNDERROR: No Module Named 'FooBar'"),
                             sv("python"), 1);
        expect((*r == "Python module FooBar is missing. Install it (e.g. "
                      "`pip install FooBar`) or check the environment."));
        r = annotate_failure(sv("Permission denied"), sv("cat"), 1);
        expect((*r == "Permission denied. Check file permissions (ls -la) or "
                      "ownership."));
        expect(!annotate_failure(sv("everything fine"), sv("ls"), 0).has_value());
        // the sample is capped at 4000 chars: a hit beyond that is invisible
        const std::string sample5000(5000, 'x');
        expect(!annotate_failure(sv(sample5000 + "command not found"), sv("x"), 1).has_value());
    };

    // ---------------------------------------------------------------------------
    // Benchmarks -- shell-safety kernels (kimix_bench contract, bench_util.h).
    // Production shape: every tool command is checked with
    // check_hardline_blocked (variants + detect), detect_hardline_command and
    // command_detection_variants run on every command in a session, and
    // base_command_name on every launched process name.  Workloads: 10k
    // diverse commands (safe + blocked + obfuscated mixes, quotes/escapes)
    // and 10k process paths.
    // ---------------------------------------------------------------------------

    "bench_detect_hardline_10k"_test = [] {
        const cmd_fixtures f = build_cmd_fixtures();
        size_t blocked = 0;
        kimix_bench::run("shell/detect_hardline_10k", [&] {
            for (const auto& c : f.cmds) {
                hardline_result r = detect_hardline_command(kimix::string_view(c));
                if (r.blocked) {
                    ++blocked;
                }
            }
        }, 10000, f.bytes);
        expect((blocked % f.e_detect == 0u));
        expect((blocked > 0u));
        kimix_bench::sink(blocked);
    };

    "bench_check_blocked_10k"_test = [] {
        const cmd_fixtures f = build_cmd_fixtures();
        size_t blocked = 0;
        kimix_bench::run("shell/check_blocked_10k", [&] {
            for (const auto& c : f.cmds) {
                hardline_result r = check_hardline_blocked(kimix::string_view(c));
                if (r.blocked) {
                    ++blocked;
                }
            }
        }, 10000, f.bytes);
        expect((blocked % f.e_check == 0u));
        expect((blocked > 0u));
        kimix_bench::sink(blocked);
    };

    "bench_variants_10k"_test = [] {
        const cmd_fixtures f = build_cmd_fixtures();
        kimix::vector<kimix::string> vout;
        size_t vtotal = 0;
        kimix_bench::run("shell/variants_10k", [&] {
            for (const auto& c : f.cmds) {
                command_detection_variants(kimix::string_view(c), vout);
                vtotal += vout.size();
            }
        }, 10000, f.bytes);
        expect((vtotal % f.e_variants == 0u));
        expect((vtotal > 0u));
        kimix_bench::sink(vtotal);
    };

    "bench_base_command_name_10k"_test = [] {
        kimix::vector<kimix::string> cmds;
        cmds.reserve(10000);
        double bytes = 0.0;
        for (size_t i = 0; i < 10000; ++i) {
            char b[80];
            std::snprintf(b, sizeof(b), "/usr/bin/kit_%04zu.exe --work %zu",
                          i, i);
            cmds.emplace_back(b);
            bytes += static_cast<double>(cmds.back().size());
        }
        size_t size_total = 0;
        kimix_bench::run("shell/base_cmd_name_10k", [&] {
            for (const auto& c : cmds) {
                size_total += base_command_name(kimix::string_view(c)).size();
            }
        }, 10000, bytes);
        // Correctness pass (unmeasured): every path maps to its expected stem.
        for (size_t i = 0; i < 10000; ++i) {
            char exp[32];
            std::snprintf(exp, sizeof(exp), "kit_%04zu", i);
            expect((base_command_name(kimix::string_view(cmds[i])) == exp));
        }
        expect((size_total >= 5u * 10000u));
        kimix_bench::sink(size_total);
    };
}

File: python/tests/test_shell_safety.py
"""Parity tests for kimix_native.tools shell-safety kernels (plan: commit
0582e09 "Study from hermes").

Compares the native kernels against the verbatim _compat mirrors over
adversarial ASCII corpora (native path) and non-ASCII commands (compat path,
equal by construction).  Gate toggled in-process via KIMIX_NATIVE_TOOLS=0.

Coverage:
- command_detection_variants: collapse/dedup/deobfuscate
- detect_hardline_command / check_hardline_blocked: all 7 rule types +
  obfuscation defeat + benign passthrough
- foreground_background_guidance: 12 patterns + quoted-span stripping
  - base_command_name: first non-assignment command word
  - interpret_exit_code / is_expected_exit: well-known exit codes and
    expected-exit classification (pure Python in the shim — no native kernel)
- annotate_failure: 4000-char sample cap, module-not-found capture
"""

import pytest

from kimix_native import tools as T


def _run_both(fn, *args, monkeypatch):
    monkeypatch.delenv("KIMIX_NATIVE_TOOLS", raising=False)
    native = fn(*args)
    monkeypatch.setenv("KIMIX_NATIVE_TOOLS", "0")
    compat = fn(*args)
    assert native == compat, f"native != compat for {args!r}:\n{native!r}\n{compat!r}"
    return native, compat


# ---------------------------------------------------------------------------
# command_detection_variants
# ---------------------------------------------------------------------------

VARIANT_CASES = [
    "rm -rf /",
    "  rm   -rf  /  ",
    r"r\m -rf /",
    "",
    "   ",
    '"rm" -rf /',
    "echo 'hi'",
    "rm -rf / && echo done",
    "RM -RF /",
    "s\u00fcdO rm -rf /",  # non-ASCII routes to compat
]


@pytest.mark.parametrize("command", VARIANT_CASES)
def test_variants_parity(command, monkeypatch):
    native, compat = _run_both(T.command_detection_variants, command, monkeypatch=monkeypatch)
    assert compat == T._compat_command_detection_variants(command)


def test_variants_goldens(monkeypatch):
    monkeypatch.delenv("KIMIX_NATIVE_TOOLS", raising=False)
    assert T.command_detection_variants("") == []
    assert T.command_detection_variants("   ") == []
    assert T.command_detection_variants("rm -rf /") == ["rm -rf /"]
    assert T.command_detection_variants(r"r\m -rf /") == [r"r\m -rf /", "rm -rf /"]
    assert T.command_detection_variants("RM -RF /") == ["RM -RF /", "rm -rf /"]


# ---------------------------------------------------------------------------
# detect_hardline_command / check_hardline_blocked
# ---------------------------------------------------------------------------

HARDLINE_CASES = [
    "",
    "   ",
    "ls -la",
    "echo hello",
    "rm -rf /",
    "rm -rf ~",
    "rm -rf $HOME",
    "rm -rf ${HOME}",
    "rm -rf /*",
    "rm -rf /.",
    "rm -r /",
    "rm -f /",
    "rm /",
    "rm -rf /tmp/build",
    "rmdir /",
    "rmdir -r /",
    "rmdir -s /",
    "rmdir -p /",
    "del /f /s /q C:\\*",
    "del /s /q C:\\",
    "del /q C:\\Windows",
    "rm -rf C:\\",
    "rm -rf c:/",
    "rm -rf C:\\*",
    r"r\m -rf /",
    "r'm' -rf /",
    "rm '-rf' /",
    "mkfs.ext4 /dev/sda1",
    "mkfs /dev/sda",
    "mkfs.fat -n USB /dev/sdb1",
    "mkfs.xfs /dev/sdc",
    "dd if=/dev/zero of=/dev/sda",
    "dd if=/dev/zero of=/dev/nvme0n1",
    "dd if=/dev/zero of=/dev/disk2",
    "dd of=/dev/rdisk1",
    "dd if=/dev/zero of=/tmp/out",
    "dd if=/dev/zero of=/dev/sd",
    "shutdown -h now",
    "reboot",
    "poweroff",
    "halt",
    "shutdown",
    ":(){ :|:& };:",
    ":(){ :& };:",
    "kill 1",
    "kill -9 1",
    "kill $PPID",
    "kill 123",
    "kill -9 123",
    "killall 1",
    "format C:",
    "format D:\\",
    "format C:\\Windows",
    "format c:",
    "sudo rm -rf /",
    "git rm -rf /",
    "echo rm -rf /",
    "rm -rf /\u00e9",  # non-ASCII routes to compat
]


@pytest.mark.parametrize("command", HARDLINE_CASES)
def test_detect_hardline_parity(command, monkeypatch):
    native, compat = _run_both(T.detect_hardline_command, command, monkeypatch=monkeypatch)
    assert compat == T._compat_detect_hardline_command(command)


@pytest.mark.parametrize("command", HARDLINE_CASES)
def test_check_hardline_parity(command, monkeypatch):
    native, compat = _run_both(T.check_hardline_blocked, command, monkeypatch=monkeypatch)
    assert compat == T._compat_check_hardline_blocked(command)


@pytest.mark.parametrize(
    "command,expected",
    [
        ("rm -rf /", "Recursive delete of protected root/home (`/`)"),
        ("rm -rf $HOME", "Recursive delete of protected root/home (`$home`)"),
        ("rm -rf ${HOME}", "Recursive delete of protected root/home (`${home}`)"),
        ("del /f /s /q C:\\*", "Recursive delete of protected root/home (`c:\\*`)"),
        ("mkfs.ext4 /dev/sda1", "Disk formatting command (`mkfs`) is blocked"),
        ("dd if=/dev/zero of=/dev/sda", "`dd` writing to a raw device is blocked"),
        ("shutdown -h now", "System `shutdown` command is blocked"),
        (":(){ :|:& };:", "Fork bomb pattern detected"),
        ("kill 1", "`kill` targeting PID 1 (or `$PPID`) is blocked"),
        ("format C:", "Windows `format` on a drive is blocked"),
    ],
)
def test_detect_hardline_goldens(command, expected, monkeypatch):
    monkeypatch.delenv("KIMIX_NATIVE_TOOLS", raising=False)
    blocked, desc = T.detect_hardline_command(command)
    assert blocked and desc == expected


def test_check_hardline_obfuscation(monkeypatch):
    monkeypatch.delenv("KIMIX_NATIVE_TOOLS", raising=False)
    for cmd in (r"r\m -rf /", "r'm' -rf /", "rm '-rf' /", r"R\M -RF /"):
        blocked, desc = T.check_hardline_blocked(cmd)
        assert blocked, cmd
        assert desc == "Recursive delete of protected root/home (`/`)", cmd
    blocked, _ = T.check_hardline_blocked("echo hello")
    assert not blocked


# ---------------------------------------------------------------------------
# foreground_background_guidance
# ---------------------------------------------------------------------------

GUIDANCE_CASES = [
    "",
    "   ",
    "ls -la",
    "npm run dev",
    "npm run start",
    "npm run build",
    "pnpm run watch",
    "yarn run serve",
    "bun run dev",
    "next dev",
    "vite",
    "nodemon app.js",
    "uvicorn app:app",
    "gunicorn app:app",
    "python -m http.server",
    "docker compose up",
    "docker-compose up",
    "echo done &",
    "echo done & ",
    "nohup python app.py",
    "setsid bash",
    "echo 'npm run dev'",
    'echo "docker compose up"',
    "grep npm run dev x",
    "npmrun dev",
    "vite \u00e9",  # non-ASCII routes to compat
]


@pytest.mark.parametrize("command", GUIDANCE_CASES)
def test_guidance_parity(command, monkeypatch):
    native, compat = _run_both(T.foreground_background_guidance, command, monkeypatch=monkeypatch)
    assert compat == T._compat_foreground_background_guidance(command)


def test_guidance_goldens(monkeypatch):
    monkeypatch.delenv("KIMIX_NATIVE_TOOLS", raising=False)
    hint = (
        "Long-running process detected. Consider mode='send' (background) + "
        "TaskOutput to avoid blocking on timeout."
    )
    assert T.foreground_background_guidance("npm run dev") == hint
    assert T.foreground_background_guidance("docker-compose up") == hint
    assert T.foreground_background_guidance("echo done &") == hint
    assert T.foreground_background_guidance("npm run build") is None
    assert T.foreground_background_guidance("echo 'npm run dev'") is None
    assert T.foreground_background_guidance("") is None


# ---------------------------------------------------------------------------
# base_command_name / interpret_exit_code
# ---------------------------------------------------------------------------

BASE_NAME_CASES = [
    "/usr/bin/grep -r foo",
    "FOO=1 git diff",
    "python -m http.server",
    "app.exe --help",
    "cmd.exe",
    "echo hi && grep foo",
    "",
    "ls",
    "   ",
    "a=b c=d",
    "python3.11 -V",
    "/x/y/tool.EXE arg",
    "ls \u00e9",  # non-ASCII routes to compat
]


@pytest.mark.parametrize("command", BASE_NAME_CASES)
def test_base_name_parity(command, monkeypatch):
    native, compat = _run_both(T.base_command_name, command, monkeypatch=monkeypatch)
    assert compat == T._compat_base_command_name(command)


def test_base_name_goldens(monkeypatch):
    monkeypatch.delenv("KIMIX_NATIVE_TOOLS", raising=False)
    assert T.base_command_name("/usr/bin/grep -r foo") == "grep"
    assert T.base_command_name("FOO=1 git diff") == "git"
    assert T.base_command_name("cmd.exe") == "cmd"
    assert T.base_command_name("/x/y/tool.EXE arg") == "tool"
    assert T.base_command_name("a=b c=d") == ""
    assert T.base_command_name("") == ""


EXIT_CODE_CASES = [
    ("grep foo", 1),
    ("grep foo", 2),
    ("grep foo", 0),
    ("grep foo", None),
    ("diff a b", 1),
    ("find . -name x", 1),
    ("test -f x", 1),
    ("[ -f x ]", 1),
    ("curl https://x", 6),
    ("curl https://x", 7),
    ("curl https://x", 22),
    ("curl https://x", 28),
    ("curl https://x", 99),
    ("git diff", 1),
    ("git diff", 2),
    ("ls", 1),
    ("", 1),
    ("   ", 1),
    ("egrep foo", 1),
    ("rg foo", 1),
    ("colordiff a b", 1),
    ("C:\\tools\\grep.exe foo", 1),
    ("python -m pytest", 5),
    ("grep \u00e9", 1),  # non-ASCII routes to compat
]


@pytest.mark.parametrize("command,code", EXIT_CODE_CASES)
def test_exit_code_python(command, code):
    # interpret_exit_code is pure Python in the shim (no native kernel), so it
    # must simply delegate to the compat mirror.
    assert T.interpret_exit_code(command, code) == T._compat_interpret_exit_code(command, code)


def test_exit_code_goldens():
    assert T.interpret_exit_code("grep foo", 1) == "No matches found (not an error)"
    assert T.interpret_exit_code("curl https://x", 6) == "Could not resolve host (DNS failure)"
    assert T.interpret_exit_code("git diff", 1).startswith("Non-zero exit")
    assert T.interpret_exit_code("grep foo", 0) is None
    assert T.interpret_exit_code("grep foo", None) is None
    # SIGPIPE inside a pipeline is the normal truncation meaning.
    assert T.interpret_exit_code("yes | head", 141) == (
        "SIGPIPE: an upstream pipeline stage was truncated "
        "(expected when piping to head/tail)"
    )


# ---------------------------------------------------------------------------
# is_expected_exit (pure Python in the shim)
# ---------------------------------------------------------------------------

EXPECTED_CASES = [
    ("grep foo", 1, True),
    ("egrep foo", 1, True),
    ("rg foo", 1, True),
    ("diff a b", 1, True),
    ("colordiff a b", 1, True),
    ("test -f x", 1, True),
    ("[ -f x ]", 1, True),
    ("find . -name x", 1, True),
    ("ls", 1, False),
    ("git merge", 1, False),
    ("grep foo", 2, False),
    ("grep foo", 0, False),
    ("grep foo", None, False),
    ("", 1, False),
    # SIGPIPE: expected only when a top-level pipeline exists.
    ("yes | head", 141, True),
    ("producer --lines | tail -n 5", 141, True),
    ("yes", 141, False),
    ("echo 'a | b'", 141, False),  # pipe inside quotes is not a pipeline
    ("a || b", 141, False),  # logical-OR is not a pipeline
]


@pytest.mark.parametrize("command,code,expected", EXPECTED_CASES)
def test_is_expected_exit(command, code, expected):
    assert T.is_expected_exit(command, code) is expected
    assert T.is_expected_exit(command, code) == T._compat_is_expected_exit(command, code)


# ---------------------------------------------------------------------------
# annotate_failure
# ---------------------------------------------------------------------------

ANNOTATE_CASES = [
    ("", "x", 1),
    ("bash: foo: command not found", "foo", 127),
    ("'foo' is not recognized as an internal or external command", "foo", 1),
    ("ls: cannot access 'x': No such file or directory", "ls", 2),
    ("ModuleNotFoundError: No module named 'requests'", "python", 1),
    ("Traceback (most recent call last):\nModuleNotFoundError: No module named 'pandas'",
     "python", 1),
    ("MODULENOTFOUNDERROR: No Module Named 'FooBar'", "python", 1),
    ("Modulenotfounderror: no module named 'abc'", "python", 1),
    ("Permission denied", "cat", 1),
    ("everything fine", "ls", 0),
    ("x" * 5000 + "command not found", "x", 1),
    ("no such file or directory", "ls", 1),
    ("caf\u00e9 command not found", "x", 1),  # non-ASCII output -> compat
]


@pytest.mark.parametrize("output,command,code", ANNOTATE_CASES)
def test_annotate_parity(output, command, code, monkeypatch):
    native, compat = _run_both(T.annotate_failure, output, command, code, monkeypatch=monkeypatch)
    assert compat == T._compat_annotate_failure(output, command, code)


def test_annotate_goldens(monkeypatch):
    monkeypatch.delenv("KIMIX_NATIVE_TOOLS", raising=False)
    r = T.annotate_failure("ModuleNotFoundError: No module named 'requests'", "python", 1)
    assert r == ("Python module requests is missing. Install it "
                 "(e.g. `pip install requests`) or check the environment.")
    r = T.annotate_failure("MODULENOTFOUNDERROR: No Module Named 'FooBar'", "python", 1)
    assert r == ("Python module FooBar is missing. Install it "
                 "(e.g. `pip install FooBar`) or check the environment.")
    assert T.annotate_failure("", "x", 1) is None
    assert T.annotate_failure("everything fine", "ls", 0) is None
